### PR TITLE
User API endpoint testing

### PR DESCRIPTION
**Pull request by maryjanebarger**
Java, TestNG, REST-Assured and Maven used to test User REST API endpoints.

### CODE REVIEW

File: api_end_end_testing/src/main/java/tds/user/api/model/RoleAssociation.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
+package tds.user.api.model;
+
+/**
+ * Created by mjbarger on 9/6/16.
```
Comment (jtreuting) [inline api_end_end_testing/src/main/java/tds/user/api/model/RoleAssociation.java]: Let's change the comment to talk about what the class represents and is used for.

File: api_end_end_testing/src/main/java/tds/user/api/model/RoleAssociation.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
+package tds.user.api.model;
```
Comment (jtreuting) [inline api_end_end_testing/src/main/java/tds/user/api/model/RoleAssociation.java]: I don't know best practices for this myself so someone please jump in.  

Does it make sense for these to live in the test package instead since they are only used there?

@millerts @jeffjohnson9046 @burzum619

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+
+        accessToken = given()
+            .contentType("application/x-www-form-urlencoded")
+            .header(header)
+            .queryParam("realm", "/sbac")
+            .formParam("client_id", "pm")
+            .formParam("client_secret", "sbac12345")
+            .formParam("grant_type", "password")
+            .formParam("password", "password")
+            .formParam("username", "prime.user@example.com")
```
Comment (jtreuting) [inline api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: We need the credentials to live outside the source code.  So even though this is our dev environment we don't want this out there for people to see.  Where is the best place for this to live?  I'm guessing a resource type file that is ignored by the .gitignore file and then we can create it locally and on the CI/Build server.

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+        }
+
+        given()
+            .contentType(ContentType.JSON)
+            .header(authHeader)
+            .body(userInfo)
+            .when()
+            .post("/rest/external/user")
+            .then()
+            .statusCode(201);
```
Comment (jtreuting) [inline api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: Let's stay consistent with the indenting on the fluent methods.  I personally like the indenting that @millerts  uses here: https://github.com/SmarterApp/TDS_StudentService/pull/1/files#diff-facd1a87945e48f2a54419311a7880a1R17

So moving when() and then() to the left at the same level as given()

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+    */
+    @Test(dependsOnMethods = "getAuthenticationCode")
+    public void createUser() {
+        email = "betsy.ross@example.com";
+
+        RestAssured.baseURI = userUri;
+
+        List<RoleAssociation> roleAssociations = new ArrayList<RoleAssociation>();
+        roleAssociations.add(new RoleAssociation("Administrator", "CLIENT", "98765"));
+        roleAssociations.add(new RoleAssociation("Administrator", "CLIENT", "98765"));
```
Comment (jtreuting) [inline api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: Duplicate roles listed.  Is there a need for 2 roles or was that by accident?  This is also done a few places below as well.

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+import org.testng.annotations.Test;
+import tds.user.api.model.UserInfo;
+import tds.user.api.model.RoleAssociation;
+
+/*
+    Authenticate user
+    Testing HTTP POST of https://sso-deployment.sbtds.org/auth/oauth2/access_token?realm=/sbac
+    and success status 200
+ */
+public class CreateUserTest extends BaseUri{
```
Comment (jtreuting) [inline api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: I think the name of this class should be something like UserApiTests

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+
+    @Test
+    public void getAuthenticationCode() {
+        RestAssured.baseURI = authenticateURI;
+
+        Header header = new Header("Content-Type", "application/x-www-form-urlencoded");
+
+        accessToken = given()
+            .contentType("application/x-www-form-urlencoded")
+            .header(header)
```
Comment (jtreuting) [inline api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: looks like header is not needed since it is set via the contentType() method.  at least it seems like that would be the case.

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+    Authenticate user
+    Testing HTTP POST of https://sso-deployment.sbtds.org/auth/oauth2/access_token?realm=/sbac
+    and success status 200
+ */
+public class CreateUserTest extends BaseUri{
+    String accessToken = null;
+    String email = null;
+
+    @Test
+    public void getAuthenticationCode() {
```
Comment (jtreuting) [inline api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: I'm not sure this feels like a test to me since we are not testing the OAuth piece but really just need the token back.  I would suggest moving this to part of the @BeforeClass method that you have and getting the access_token there.  Then each Test will have access to it since it is the base class.

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+            .statusCode(200)
+            .extract()
+            .path("access_token");
+    }
+
+    /*
+        Create user
+        Testing HTTP POST of /rest/external/user, 201 success item created
+    */
+    @Test(dependsOnMethods = "getAuthenticationCode")
```
Comment (jtreuting) [inline api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: If we restructure like above, then we can get rid of the dependsOnMethods and each test would be self-sufficient at that point instead of relying on other tests.

File: api_end_end_testing/src/main/java/tds/user/api/model/RoleAssociation.java (hunk 2 above)
Comment (millerts) [reply api_end_end_testing/src/main/java/tds/user/api/model/RoleAssociation.java]: I do not think the class needs to be in a package named `test`.  That being said I would recommend that we move it into `test/java` rather than `main/java`.  The reason I say this is because the code in `main/java` doesn't really serve any purpose outside of running the integration tests.

If this code was executed outside of maven (like an executable jar) then I'd say it should be in main.

Comment (jtreuting) [reply api_end_end_testing/src/main/java/tds/user/api/model/RoleAssociation.java]: Sorry, that is what I meant.  Inside "test" instead of inside "main."  Bad wording on my part.

And no, I think Maven will be the one always running it.

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java (hunk 3 above)
Comment (millerts) [reply api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: The simplest way we could do this is to pass in arguments via the command line.  For example:

`mvn test -Dserver="http://localhost:8080/"` and then the @BeforeClass could get the system property by calling `System.getProperty("server")` and if it isn't set it could immediately fail the test with an error explaining why.

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+        List<RoleAssociation> roleAssociations = new ArrayList<RoleAssociation>();
+        roleAssociations.add(new RoleAssociation("Administrator", "CLIENT", "98765"));
+        roleAssociations.add(new RoleAssociation("Administrator", "CLIENT", "98765"));
+
+        Header authHeader = new Header("Authorization", "Bearer " + accessToken);
+
+        UserInfo userInfo = new UserInfo(email, "Betsy", "Ross", "800-332-4747", roleAssociations);
+
+        ObjectMapper mapper = new ObjectMapper();
+        String jsonInString = null;
```
Comment (millerts) [inline api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: Lines 67-74 seem to be debug code or left over code from an earlier strategy to test the endpoints.  I suggest that you can remove them because jsonInString doesn't seem to be used anymore in the code starting at line 76.

Suggest removing all of that type of code unless there is a compelling reason to leave it.

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+        System.out.println("endpoint: " + "/rest/external/user/" + email + "/details");
+
+        given()
+            .contentType(ContentType.JSON)
+            .accept(ContentType.JSON)
+            .header(authHeader)
+            .when()
+            .get("/rest/external/user/" + email + "/details")
+            .then()
+            .statusCode(200);
```
Comment (millerts) [inline api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: Since this endpoint returns some data I think we should test to make sure it contains the JSON we expect.  We do this in other projects using [hamcrest](https://code.google.com/archive/p/hamcrest/wikis/Tutorial.wiki]).  For example, this endpoint returns:

``` json
{
  "email" : "betsy.ross@example.com",
  "firstName" : "Betsy",
  "lastName" : "Carson",
  "phoneNumber" : "800-332-4747",
  "roleAssociations" : [ {
    "role" : "Administrator",
    "level" : "CLIENT",
    "entityId" : "98765"
  }
```

Taking that into account we can test the email by tweaking your RestAssured code to like this:

``` java
        given()
            .contentType(ContentType.JSON)
            .accept(ContentType.JSON)
            .header(authHeader)
        .when()
            .get("/rest/external/user/" + email + "/details")
        .then()
            .statusCode(200)
            .body("email", equalTo("betsy.ross@example.com"));
```

That .body("email"...) line is inspecting the JSON response.  You can chain them along to check for multiple things in the JSON body payload.  Here is an [example](https://github.com/SmarterApp/TDS_SessionService/blob/develop/service/src/test/java/tds/session/web/endpoints/SessionControllerIntegrationTests.java#L32).

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java (hunk 3 above)
Comment (millerts) [reply api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: Please ignore my suggestion.  Though it will work in this particular case it will quickly become cumbersome when we have more than one service being tested.  I think what you have plus what @jtreuting mentioned is a much better solution.

File: api_end_end_testing/src/test/java/tds/user/api/tests/BaseUri.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+
+        Properties prop = new Properties();
+        InputStream inputStream = getClass().getClassLoader().getResourceAsStream("config.properties");
+
+        try {
+            prop.load(inputStream);
+            System.out.println("auth prop " + prop.getProperty("authenticateURI"));
+        }
+        catch (IOException e) {
+            e.printStackTrace(System.out);
```
Comment (millerts) [inline api_end_end_testing/src/test/java/tds/user/api/tests/BaseUri.java]: Rather than printing out a stacktrace what if instead we did `fail(e.getMessage())`?  That way the test run would just end right then.  And if that doesn't work I suggest doing going with `throw new RuntimeException(e)`.  Either way that would result in the test catastrophically failing which is what I think we want since the tests will all fail anyway without the properties file.

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java (hunk 3 above)
Comment (jeffjohnson9046) [reply api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: I agree w/JT - keeping credentials in an external file that is ignored via `.gitignore` is probably the best bet.  We're doing something vaguely similar with the Docker setup.  While I don't think the exact same technique will work, I think it might be a good starting point.

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java [hunk 13]
```diff
@@ -0,0 +1,247 @@
...
+            .post("/rest/external/user")
+            .then()
+            .statusCode(204);
+    }
+
+    /*
+        Create invalid user
+        Testing HTTP POST of /rest/external/user, 400 bad request
+    */
+    @Test(dependsOnMethods = "updateUser")
```
Comment (jeffjohnson9046) [inline api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: Maybe it's just me, but I'm finding the `dependsOnMethods` to be a bit hard to follow.  I understand that the tests might need to be executed in a particular order, and that JUnit doesn't guarantee execution order (by design).  As complexity of workflows increases, I think this could become brittle and/or hard to follow.  Also, doesn't this make it a bit harder to run the tests individually?  For instance, what happens if I try to run the `updateUser()` test?  Will the `dependsOnMethods` attribute cause the dependent test to be executed?

File: api_end_end_testing/src/main/java/tds/user/api/model/RoleAssociation.java (hunk 1 above)
Comment (maryjanebarger) [reply api_end_end_testing/src/main/java/tds/user/api/model/RoleAssociation.java]: Done.

File: api_end_end_testing/src/main/java/tds/user/api/model/RoleAssociation.java (hunk 2 above)
Comment (maryjanebarger) [reply api_end_end_testing/src/main/java/tds/user/api/model/RoleAssociation.java]: Okay. Followed standard Maven directory layout where test directories miror src directories. But will change if majority prefers.  https://maven.apache.org/guides/introduction/introduction-to-the-standard-directory-layout.html

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java (hunk 4 above)
Comment (maryjanebarger) [reply api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: IntelliJ reformatted my code.

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java (hunk 5 above)
Comment (maryjanebarger) [reply api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: Meant to create a list of different roles. Will fix.

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java (hunk 6 above)
Comment (maryjanebarger) [reply api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: Done

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java (hunk 7 above)
Comment (maryjanebarger) [reply api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: Removed

File: api_end_end_testing/src/main/java/tds/user/api/model/RoleAssociation.java (hunk 2 above)
Comment (millerts) [reply api_end_end_testing/src/main/java/tds/user/api/model/RoleAssociation.java]: Normally I would have the code residing in the structure present in this project, but this project is unique as in it will most likely only run via Maven's test lifecycle.  So I would want the Maven directory structure to remain but the `src/main` folders will be empty with everything existing within `src/test/java`.

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java (hunk 3 above)
Comment (millerts) [reply api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: To verify what I mentioned at stand up today I modified the base class to look for a system property rather than looking for a file on the filepath.  The example can be seen [here](https://github.com/SmarterApp/TDS_TestDeliverySystemRegressionTest/blob/configFileExample/api_end_end_testing/src/test/java/tds/user/api/tests/BaseUri.java)

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java (hunk 4 above)
Comment (millerts) [reply api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: Unfortunately, I haven't figured out how to tell IntelliJ not to move all indents over the necessary spaces so I have to do it manually :(

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+            e.printStackTrace();
+        }
+
+        given()
+            .contentType(ContentType.JSON)
+            .header(authHeader)
+            .body(userInfo)
+            .when()
+            .post("/rest/external/user")
+            .then()
```
Comment (millerts) [inline api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: I spoke a bit with @hbgwhite, and he confirmed that the user API is sending a "location" header with the link to use to get the newly created user (seen [here](https://github.com/SmarterApp/TDS_AdministrationAndRegistrationTools/pull/1/files#diff-3a41820a8dc9f59e2234ee35fa55eac7R84)).  I believe you should be able to verify that the header has that populated when this returns with the 201 status code by chaining `.header("location", endsWith("/rest/external/user/betsy.ross@example.com/details"))`.  Or at least that is what I think should come back in the header based on the linked code.

File: api_end_end_testing/src/test/java/tds/user/api/tests/BaseUri.java [hunk 15]
```diff
@@ -1,31 +1,70 @@
...
     public void init() {
+        // Get properties needed for authentication
+/*
+        String configFileLocation = System.getProperty("config");
 
+        if (configFileLocation == null || configFileLocation.isEmpty()) {
+            throw new IllegalArgumentException("config file location system property is required (mvn test -Dconfig=<fileLocation>");
+        }
+*/
+        String configFileLocation = "/Users/mjbarger/config.properties";
```
Comment (jeffjohnson9046) [inline api_end_end_testing/src/test/java/tds/user/api/tests/BaseUri.java]: Instead of hard-coding the path to the `config.properties` file, would it be possible to use an annotation or something to say "this is the `properties` file name"?  What I'm getting at is instead of saying the path to the config file is `/Users/mjbarger/config.properties` maybe say the path is `classpath:/config.properties`.  It looks like this might've worked:  `getClass().getClassLoader().getResourceAsStream("config.properties")`, but I could be totally wrong and/or missing something.

File: api_end_end_testing/src/main/java/tds/user/api/model/RoleAssociation.java (hunk 2 above)
Comment (maryjanebarger) [reply api_end_end_testing/src/main/java/tds/user/api/model/RoleAssociation.java]: Moved it. Done

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java (hunk 3 above)
Comment (maryjanebarger) [reply api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: Done.

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java (hunk 4 above)
Comment (maryjanebarger) [reply api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: Will try but it was IntelliJ that left aligned everything.

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java (hunk 5 above)
Comment (maryjanebarger) [reply api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: Fixed.

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java (hunk 8 above)
Comment (maryjanebarger) [reply api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: Moved it to my BaseUri class

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java (hunk 9 above)
Comment (maryjanebarger) [reply api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: Done

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java (hunk 10 above)
Comment (maryjanebarger) [reply api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: Removed it

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java (hunk 11 above)
Comment (maryjanebarger) [reply api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: Done

File: api_end_end_testing/src/test/java/tds/user/api/tests/BaseUri.java (hunk 12 above)
Comment (maryjanebarger) [reply api_end_end_testing/src/test/java/tds/user/api/tests/BaseUri.java]: Code has been changed.

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java (hunk 13 above)
Comment (maryjanebarger) [reply api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: Eliminated

File: api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java (hunk 14 above)
Comment (maryjanebarger) [reply api_end_end_testing/src/test/java/tds/user/api/tests/CreateUserTest.java]: Done

File: api_end_end_testing/src/test/java/tds/user/api/tests/BaseUri.java (hunk 15 above)
Comment (maryjanebarger) [reply api_end_end_testing/src/test/java/tds/user/api/tests/BaseUri.java]: Handled.

File: api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java [hunk 16]
```diff
@@ -0,0 +1,359 @@
...
+            .contentType(ContentType.JSON)
+            .accept(ContentType.JSON)
+            .header(authHeader)
+        .when()
+            .get(uriLocation + "/" + userEmail + "/details")
+        .then()
+            .statusCode(200)
+            .body("firstName", is(userInfo.getFirstName()))
+            .body("lastName", is(userInfo.getLastName()))
+            .body("phoneNumber", is(userInfo.getPhoneNumber()));
```
Comment (jtreuting) [inline api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java]: I think we should check the role associations as well to make sure it matches what was passed in.

File: api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java [hunk 17]
```diff
@@ -0,0 +1,359 @@
...
+
+        // Execute a DELETE to delete the user
+        deleteExistingUser(randomUserEmail);
+
+        // Execute a GET by email to validate user is deleted
+        given()
```
Comment (jtreuting) [inline api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java]: I think we could include this in the deleteExistingUser() and have it always be there.  After deleting a user the only real way to tell it was successful is to do this GET and check for a 404.  So I say let's always do this in the delete method.

File: api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java [hunk 18]
```diff
@@ -0,0 +1,359 @@
...
+            .contentType(ContentType.JSON)
+            .accept(ContentType.JSON)
+            .header(authHeader)
+        .when()
+            .get(uriLocation + "/" + randomUserEmail + "/details")
+        .then()
+            .statusCode(200)
+            .body("firstName", is(userInfo.getFirstName()))
+            .body("lastName", is(userInfo.getLastName()))
+            .body("phoneNumber", is(userInfo.getPhoneNumber()));
```
Comment (jtreuting) [inline api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java:144]: Add check to make sure role associations were updated appropriately.

File: api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java [hunk 19]
```diff
@@ -0,0 +1,359 @@
...
+
+        // Execute a DELETE to delete the user
+        deleteExistingUser(randomUserEmail);
+
+        // Execute a GET by email to validate user is deleted
+        given()
```
Comment (jtreuting) [inline api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java]: same comment as above

File: api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java [hunk 20]
```diff
@@ -0,0 +1,359 @@
...
+        .when()
+            .post(uriLocation)
+        .then()
+            .statusCode(204);
+
+        // Execute a DELETE to delete the user
+        deleteExistingUser(randomUserEmail);
+
+        // Execute a GET by email to validate user is deleted
+        given()
```
Comment (jtreuting) [inline api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java]: ditto

File: api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java [hunk 21]
```diff
@@ -0,0 +1,359 @@
...
+     * Test of Create User with invalid email, HTTP POST of /rest/external/user, 400 bad request
+     */
+    @Test
+    public void shouldNotCreateUserWithInvalidEmail() {
+        List<RoleAssociation> roleAssociations = new ArrayList<RoleAssociation>();
+        roleAssociations.add(new RoleAssociation("Administrator", "CLIENT", "19037"));
+
+        UserInfo userInfo = new UserInfo("bademail", "Judy", "Bloom", "808-443-1199", roleAssociations);
+
+        given()
```
Comment (jtreuting) [inline api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java:195]: IntelliJ auto-corrected your tabs in a couple places I think.

**Review by jtreuting — COMMENTED** [review]
Couple small things but this is looking really good.  A lot more coverage than before.

File: api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java [hunk 22]
```diff
@@ -0,0 +1,359 @@
...
+        List<RoleAssociation> roleAssociations = new ArrayList<RoleAssociation>();
+        roleAssociations.add(new RoleAssociation("Administrator", "CLIENT", "55342"));
+
+        UserInfo userInfo = new UserInfo(randomUserEmail, "", "Dodge", "714-228-4848", roleAssociations);
+
+        given()
+            .contentType(ContentType.JSON)
+            .header(authHeader)
+            .body(userInfo)
+            .when()
```
Comment (millerts) [inline api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java]: Can we format this manually to make it look pretty?   @jtreuting , I couldn't find a way to make IntelliJ do the formatting by default.  I always have to do it manually.  The thing that stinks about it is that if you run format code in IntelliJ it'll push it back.

File: regression/README.md [hunk 23]
```diff
@@ -0,0 +1,58 @@
+# Welcome to the TDS Regression Test Suite Application
```
Comment (millerts) [inline regression/README.md:1]: We should have a similar README.md in the api regression project explaining how to run the tests in Maven and IntelliJ.  We should specify the property keys that need to be in the file.

File: api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java (hunk 22 above)
Comment (jeffjohnson9046) [reply api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java]: That's what I always have to do; format these manually.  It's a drag, but I think it's worth the effort for readability.

File: api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java [hunk 24]
```diff
@@ -0,0 +1,359 @@
...
+
+         given()
+            .contentType(ContentType.JSON)
+            .header(authHeader)
+            .body(userInfo)
+        .when()
+            .post(uriLocation)
+        .then()
+            .statusCode(400)
+         .body("messages.lastName[0]", equalTo("User Last Name is required"));
```
Comment (jeffjohnson9046) [inline api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java]: Should be indented at the same level as `statusCode()`. Along with`stausCode`, `body` is part of the response that's being evaluated.

File: api_end_end_testing/src/test/java/tds/user/api/tests/BaseUri.java (hunk 15 above)
Comment (jeffjohnson9046) [reply api_end_end_testing/src/test/java/tds/user/api/tests/BaseUri.java]: My concern here is on line 33:  `String configFileLocation = "/Users/mjbarger/config.properties";`.  I do not have a `/Users/mjbarger` directory on my computer anywhere.  If I want to run the regression tests, I have to do [at least] one of two things before I can run the regression tests on my machine:
- Create a `/Users/mjbarger` directory and put the `config.properties` file there
- Update this file to point it to where my version of the `config.properties` file is

In my opinion, neither of those are desirable.  I think it might be worth trying the following:
- Add a `resources` directory under `/src/test` at the same level as the `java` directory (so you will have `/src/test/java` and `/src/test/resources`)
- Put the `config.properties` file in the resources directory.
  - You might have to tell IntelliJ that the `resources` directory you created is in fact a "resources" - type directory.  I don't remember how to do that off the top of my head, but I know I've done it.

Once you do the steps above, the `resources` should be in the`classpath`, so all you'll have to do is say `FileInputStream("config.properties")` or something like that.  Maybe check out this link for a better explanation:  http://www.techbeamers.com/read-data-properties-file-using-testng-framework/

Comment (maryjanebarger) [reply api_end_end_testing/src/test/java/tds/user/api/tests/BaseUri.java]: The code referencing my /Users/mjbarger directory no longer exists.

File: api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java (hunk 17 above)
Comment (maryjanebarger) [reply api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java]: I pulled the GET out of the deleteExisttingUser() method to help with debugging the delete bug. I will put that code back into the delete method.

File: api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java (hunk 19 above)
Comment (maryjanebarger) [reply api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java]: Same as my comment above as well.

File: api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java (hunk 20 above)
Comment (maryjanebarger) [reply api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java]: Ditto as well.

File: api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java (hunk 22 above)
Comment (maryjanebarger) [reply api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java]: I have been formatting manually but can't control the auto changes made by IntelliJ

File: api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java (hunk 24 above)
Comment (maryjanebarger) [reply api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java]: I have been formatting manually but can't control the auto changes made by IntelliJ. I will go through and fix indenting manually again before the next commit.

File: api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java (hunk 22 above)
Comment (millerts) [reply api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java]: It may be a setting you have configured in IntelliJ.  Prior to commit there are a bunch of checkboxes on the right.  If the format code one is checked then it'll do this to you on commit.  Or if you hit option + command + enter.

It isn't ideal.

**Comment by millerts** [comment]
Approve the pull request.  Changes look great.

File: api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java (hunk 16 above)
Comment (maryjanebarger) [reply api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java]: Done

File: api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java (hunk 18 above)
Comment (maryjanebarger) [reply api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java:144]: Done

File: api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java (hunk 21 above)
Comment (maryjanebarger) [reply api_end_end_testing/src/test/java/tds/user/api/tests/UserApiTests.java:195]: Fixed
